pkg/image/importer/importer.go

```diff
@@ -0,0 +1,768 @@
+				if tag.Err != nil {
+					setImageImportStatus(isi, index, tag.Err)
+					continue
+				}
+				copied := *tag.Image
+				image := &isi.Status.Images[index]
+				ref := repo.Ref
+				ref.Tag, ref.ID = tag.Name, copied.Name
+				copied.DockerImageReference = ref.MostSpecific().Exact()
+				image.Tag = tag.Name
+				image.Image = &copied
+				image.Status.Status = unversioned.StatusSuccess
+			}
+		}
+		for _, digest := range repo.Digests {
+			j := manifestKey{repositoryKey: key}
+			j.value = digest.Name
+			if digest.Image != nil {
+				cache[j] = digest.Image
+			}
+			for _, index := range ids[j] {
+				if digest.Err != nil {
+					setImageImportStatus(isi, index, digest.Err)
+					continue
+				}
+				image := &isi.Status.Images[index]
+				copied := *digest.Image
+				ref := repo.Ref
+				ref.Tag, ref.ID = "", copied.Name
+				copied.DockerImageReference = ref.MostSpecific().Exact()
+				image.Image = &copied
+				image.Status.Status = unversioned.StatusSuccess
+			}
+		}
+	}
+}
+
+// importFromRepository imports the repository named on the ImageStreamImport, if any, importing up to maximumTags, and reporting
+// status on each image that is attempted to be imported. If the repository cannot be found or tags cannot be retrieved, the repository
+// status field is set.
+func importFromRepository(ctx *AuthContext, isi *api.ImageStreamImport, maximumTags int, cache map[manifestKey]*api.Image, limiter util.RateLimiter) {
+	if isi.Spec.Repository == nil {
+		return
+	}
+	isi.Status.Repository = &api.RepositoryImportStatus{}
+	status := isi.Status.Repository
+
+	spec := isi.Spec.Repository
+	from := spec.From
+	if from.Kind != "DockerImage" {
+		return
+	}
+	ref, err := api.ParseDockerImageReference(from.Name)
+	if err != nil {
+		status.Status = invalidStatus("", fielderrors.NewFieldInvalid("from.name", from.Name, fmt.Sprintf("invalid name: %v", err)))
+		return
+	}
+	defaultRef := ref.DockerClientDefaults()
+	repoName := defaultRef.RepositoryName()
+	registryURL := defaultRef.RegistryURL()
+
+	key := repositoryKey{url: *registryURL, name: repoName}
+	repo := &importRepository{
+		Ref:         ref,
+		Registry:    &key.url,
+		Name:        key.name,
+		Insecure:    spec.ImportPolicy.Insecure,
+		MaximumTags: maximumTags,
+	}
+	importRepositoryFromDocker(ctx, repo, limiter)
+
+	if repo.Err != nil {
+		status.Status = imageImportStatus(repo.Err, "", "repository")
+		return
+	}
+
+	additional := []string{}
+	tagKey := manifestKey{repositoryKey: key}
+	for _, s := range repo.AdditionalTags {
+		tagKey.value = s
+		if image, ok := cache[tagKey]; ok {
+			repo.Tags = append(repo.Tags, importTag{
+				Name:  s,
+				Image: image,
+			})
+		} else {
+			additional = append(additional, s)
+		}
+	}
+	status.AdditionalTags = additional
+
+	failures := false
+	status.Status.Status = unversioned.StatusSuccess
+	status.Images = make([]api.ImageImportStatus, len(repo.Tags))
+	for i, tag := range repo.Tags {
+		if tag.Err != nil {
```

Comment:
Not sure what you're asking.

On Thu, Jan 21, 2016 at 2:48 PM, David Eads <notifications@github.com>
wrote:

> In pkg/image/importer/importer.go
> <https://github.com/openshift/origin/pull/6288#discussion_r50452886>:
>
> > +		if image, ok := cache[tagKey]; ok {
> > +			repo.Tags = append(repo.Tags, importTag{
> > +				Name:  s,
> > +				Image: image,
> > +			})
> > +		} else {
> > +			additional = append(additional, s)
> > +		}
> > +	}
> > +	status.AdditionalTags = additional
> > +
> > +	failures := false
> > +	status.Status.Status = unversioned.StatusSuccess
> > +	status.Images = make([]api.ImageImportStatus, len(repo.Tags))
> > +	for i, tag := range repo.Tags {
> > +		if tag.Err != nil {
>
> What prevents you from adding a "bonus" tag from the cache that was
> errored? It would be really frustrating if I should have succeeded and the
> "bonus" tag screwed me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6288/files#r50452886>.
>